### PR TITLE
Rewrite inventory&wield mesh

### PR DESCRIPTION
**Pull request by numberZero**
screenshot_20180103_223722

Currently it changes the behavior a bit, that may or may not be desirable:

* Plantlike is not handled specially anymore
* Wield shader is not used
* Inventory images are flat if animation is disabled (because you couldn’t see the edges anyway)
* Scale may be different
* ...maybe something else.

### CODE REVIEW

**Comment by juhdanad** [comment]
I know that currently the wield mesh and item mesh generating code is very messy, there is much need for cleanup.
I think this PR should be reviewed once #6803 is merged to reduce the number of changed lines.

**Comment by paramat** [comment]
> Inventory images are flat if animation is disabled (because you couldn’t see the edges anyway)

Good idea.

**Comment by nerzhul** [comment]
just an organization fix: please push meshgen folder in client folder, like rendering engine. I haven't review other parts

**Comment by rubenwardy** [comment]
I like the new code. What needs to be done to remove [WIP]? Any remaining bugs?

**Comment by numberZero** [comment]
@rubenwardy As suggested by @juhdanad, this PR is frozen until #6803 is merged.

**Comment by numberZero** [comment]
Updated. Now this should contain refactor only, with no significant behavioral changes.

**Comment by paramat** [comment]
So #6803 doesn't have to be merged first?
Is this still WIP?

**Comment by numberZero** [comment]
Waiting for testing and reviewing)

**Comment by ClobberXD** [comment]
There still seems to be code-style issues...

File: src/client.h [hunk 1]
```diff
@@ -370,6 +372,7 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
 	ICraftDefManager* getCraftDefManager() override;
 	ITextureSource* getTextureSource();
 	virtual IShaderSource* getShaderSource();
+	IItemMeshSource* getItemMeshSource() { return m_imsrc.get(); }
```
Comment (SmallJoker) [inline src/client.h]: `IItemMeshSource *getItemMeshSource`

File: src/client/meshgen/extrusion_mesh_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+			video::S3DVertex(+r,y0,-r, 0,-1,0, c, 1,tex0),
+			video::S3DVertex(+r,y0,+r, 0,-1,0, c, 1,tex1),
+			video::S3DVertex(-r,y0,+r, 0,-1,0, c, 0,tex1),
+			// y+
+			video::S3DVertex(-r,y1,-r, 0,+1,0, c, 0,tex0),
+			video::S3DVertex(-r,y1,+r, 0,+1,0, c, 0,tex1),
+			video::S3DVertex(+r,y1,+r, 0,+1,0, c, 1,tex1),
+			video::S3DVertex(+r,y1,-r, 0,+1,0, c, 1,tex0),
+		};
+		u16 indices[12] = {0,1,2,2,3,0,4,5,6,6,7,4};
```
Comment (SmallJoker) [inline src/client/meshgen/extrusion_mesh_cache.cpp]: Couldn't this array be moved to the top of the function as static array? It's used multiple times in the loops.

File: src/client/meshgen/extrusion_mesh_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	return mesh;
+}
+
+ExtrusionMeshCache::ExtrusionMeshCache()
+{
+	for (int resolution = MIN_EXTRUSION_MESH_RESOLUTION;
+			resolution <= MAX_EXTRUSION_MESH_RESOLUTION;
+			resolution *= 2) {
+		m_extrusion_meshes[resolution] =
+				::createExtrusionMesh(resolution, resolution);
```
Comment (SmallJoker) [inline src/client/meshgen/extrusion_mesh_cache.cpp]: Superfluous `::`?

File: src/client/meshgen/item.cpp [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+void ItemMeshGenerator::createInventoryMesh()
+{
+	static thread_local bool extrude_inv =
+			g_settings->getBool("inventory_items_animations");
+	// clang-format off
+	// TODO: add custom mesh support
+	tryImage(def->inventory_image, def->inventory_overlay, extrude_inv)
+	|| tryNode();
```
Comment (SmallJoker) [inline src/client/meshgen/item.cpp]: Belong to the line above. Needs an additional indent.

File: src/client/meshgen/item.cpp [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+	rotateMeshXZby(mesh, -45);
+	rotateMeshYZby(mesh, -30);
+}
+
+void ItemMeshGenerator::createWieldMesh()
+{
+	// clang-format off
+	tryImage(def->wield_image, def->wield_overlay, true)
+	|| tryNode()
+	|| tryImage(def->inventory_image, def->inventory_overlay, true);
```
Comment (SmallJoker) [inline src/client/meshgen/item.cpp]: These here too

File: src/mesh.h [hunk 6]
```diff
@@ -35,7 +35,7 @@ void applyFacesShading(video::SColor &color, const v3f &normal);
 	The resulting mesh has 6 materials (up, down, right, left, back, front)
 	which must be defined by the caller.
 */
-scene::IAnimatedMesh* createCubeMesh(v3f scale);
+scene::SMesh* createCubeMesh(v3f scale);
```
Comment (SmallJoker) [inline src/mesh.h]: `SMesh *createCubeMesh`

File: src/client/meshgen/item.cpp [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+	scene::SMesh *mesh = nullptr;
+	std::vector<ItemPartColor> colors;
+
+	bool tryImage(const std::string &image, const std::string &overlay, bool extrude);
+	bool tryNode();
+	void createSpecialNodeMesh();
+	void postProcessMesh();
+	void postProcessNodeMesh();
+
+public:
```
Comment (SmallJoker) [inline src/client/meshgen/item.cpp]: Could you please move these public functions to the top of the class definition? Many (/all?) classes are formatted that way.

File: src/mesh.cpp [hunk 8]
```diff
@@ -61,7 +61,7 @@ void applyFacesShading(video::SColor &color, const v3f &normal)
 		applyShadeFactor(color, 0.670820f * x2 + 1.000000f * y2 + 0.836660f * z2);
 }
 
-scene::IAnimatedMesh* createCubeMesh(v3f scale)
+scene::SMesh* createCubeMesh(v3f scale)
```
Comment (SmallJoker) [inline src/mesh.cpp]: `SMesh *createCubeMesh`

File: src/client/meshgen/extrusion_mesh_cache.cpp [hunk 9]
```diff
@@ -0,0 +1,158 @@
+#include "extrusion_mesh_cache.h"
```
Comment (SmallJoker) [inline src/client/meshgen/extrusion_mesh_cache.cpp]: Also: Missing copyright header

File: src/client/meshgen/extrusion_mesh_cache.cpp (hunk 3 above)
Comment (numberZero) [reply src/client/meshgen/extrusion_mesh_cache.cpp]: no, it means global scope. It’s probably better to rename the function instead, though (currently there is a method with the same name).

File: src/client/meshgen/extrusion_mesh_cache.cpp (hunk 2 above)
Comment (numberZero) [reply src/client/meshgen/extrusion_mesh_cache.cpp]: No, some values differ on each iteration.

File: src/client/meshgen/extrusion_mesh_cache.cpp (hunk 9 above)
Comment (numberZero) [reply src/client/meshgen/extrusion_mesh_cache.cpp]: fixed

**Comment by paramat** [comment]
Assuming no longer WIP.

**Comment by ClobberXD** [comment]
Rebase needed to fix merge conflict

**Comment by paramat** [comment]
@ClobberXD thanks for your good intentions but there's no need to comment about rebasing, that's obvious already from the thread, and we end up adding a label anyway.
I also prefer to not be too demanding about rebases as they're not necessary every time a conflict arises.

**Comment by SmallJoker** [comment]
Rebased.

**Comment by numberZero** [comment]
This is probably totally obsolete (I hope so actually), and I definitely don’t plan to work on this dusty code. If refactoring is still needed in the area, that’s for another PR.

**Pull request closed without merging**